e2e/aws/eks_test.go

```diff
@@ -0,0 +1,281 @@
+	// discoveredClusterNameEnv is the environment variable that specifies the name of the EKS cluster
+	// that will be created by Teleport Discovery Service.
+	discoveredClusterNameEnv = "DISCOVERED_CLUSTER_NAME"
+)
+
+// checkRequiredEnvVars ensures that the required environment variables are set.
+func checkRequiredEnvVars(t *testing.T) {
+	require.NotEmpty(t, os.Getenv(awsRegionEnv), "AWS_REGION environment variable must be set")
+	require.NotEmpty(t, os.Getenv(kubernetesServiceAssumeRoleEnv), "KUBERNETES_SERVICE_ASSUME_ROLE environment variable must be set")
+	require.NotEmpty(t, os.Getenv(discoveryServiceAssumeRoleEnv), "DISCOVERY_SERVICE_ASSUME_ROLE environment variable must be set")
+	require.NotEmpty(t, os.Getenv(discoveredClusterNameEnv), "DISCOVERED_CLUSTER_NAME environment variable must be set")
+}
+
+func awsEKSDiscovery(t *testing.T) {
+	t.Parallel()
+	// ensures that the required environment variables are set.
+	checkRequiredEnvVars(t)
+
+	// Create the CA authority that will be used in Auth.
+	priv, pub, err := testauthority.New().GenerateKeyPair()
+	require.NoError(t, err)
+	const (
+		host   = helpers.Host
+		site   = helpers.Site
+		hostID = helpers.HostID
+	)
+	log := utils.NewLoggerForTests()
+
+	me, err := user.Current()
+	require.NoError(t, err)
+
+	teleport := helpers.NewInstance(t, helpers.InstanceConfig{
+		ClusterName: site,
+		HostID:      host,
+		NodeName:    host,
+		Priv:        priv,
+		Pub:         pub,
+		Log:         log,
+	})
+	var (
+		// Username used for the test.
+		username = me.Username
+		// kubernetes groups and users used for the test.
+		// discovery-ci-eks
+		kubeGroups = []string{kube.TestImpersonationGroup}
+		kubeUsers  = []string{"alice@example.com"}
+	)
+
+	// Create a new role with full access to all resources.
+	role, err := types.NewRole(
+		"kubemaster",
+		types.RoleSpecV6{
+			Allow: types.RoleConditions{
+				KubeGroups: kubeGroups,
+				KubeUsers:  kubeUsers,
+				KubernetesLabels: types.Labels{
+					types.Wildcard: {types.Wildcard},
+				},
+				KubernetesResources: []types.KubernetesResource{
+					{
+						Kind: types.Wildcard, Name: types.Wildcard, Namespace: types.Wildcard, Verbs: []string{types.Wildcard},
+					},
+				},
+			},
+		},
+	)
+	require.NoError(t, err)
+
+	// Create a new user with the role created above.
+	teleport.AddUserWithRole(username, role)
+	// Create a new teleport instance with the auth server.
+	err = teleport.CreateEx(t, nil, newTeleportConfig(t, log))
+	require.NoError(t, err)
+	// Start the teleport instance and wait for it to be ready.
+	err = teleport.Start()
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		require.NoError(t, teleport.StopAll())
+	})
+
+	// Get the auth server.
+	authC := teleport.Process.GetAuthServer()
+	// Wait for the discovery service to discover the cluster and create a
+	// KubernetesCluster resource.
+	// Discovery service will scan the AWS account each 5 minutes.
```

Comment:
How long does this test take on average? Is the 5 minute discovery service interval and 10 minute retry interval significant here, or just some configuration? I.e. I'm wondering will this test actually wait for 5 minutes to discover stuff in some circumstances, or will it be fast? Should we expose some hidden configuration parameter (or env var) to tweak the rescan interval and set it to like 5 seconds in tests?

Reply:
The test takes less than 10 seconds to run once the test code is compiled. 

It doesn't wait 5min for the first discovery but I use a larger timeout to protect for cases where the first discovery fails to create the kube cluster in the auth server giving it the chance to retry.


We can go full YOLO and fail if the first discovery loop fails 😂

Reply:
d890f37729b2c246f9f8d12423e785b203420238 reduces the default poll period to improve test speed